compiler/packages/snap/src/sprout/evaluator.ts

```diff
@@ -60,38 +60,55 @@ const ExportSchema = z.object({
   FIXTURE_ENTRYPOINT: EntrypointSchema,
 });
 
+const NO_ERROR_SENTINEL = Symbol();
 /**
  * Wraps WrapperTestComponent in an error boundary to simplify re-rendering
  * when an exception is thrown.
  * A simpler alternative may be to re-mount test components manually.
  */
 class WrapperTestComponentWithErrorBoundary extends React.Component<
   {fn: any; params: Array<any>},
-  {hasError: boolean; error: any}
+  {errorFromLastRender: any}
 > {
-  propsErrorMap: MutableRefObject<Map<any, any>>;
+  /**
+   * Limit retries of the child component by caching seen errors.
+   */
+  propsErrorMap: Map<any, any>;
+  lastProps: any | null;
+  // lastProps: object | null;
   constructor(props: any) {
     super(props);
-    this.state = {hasError: false, error: null};
-    this.propsErrorMap = React.createRef() as MutableRefObject<Map<any, any>>;
-    this.propsErrorMap.current = new Map();
+    this.lastProps = null;
+    this.propsErrorMap = new Map<any, any>();
+    this.state = {
+      errorFromLastRender: NO_ERROR_SENTINEL,
+    };
   }
   static getDerivedStateFromError(error: any) {
-    return {hasError: true, error: error};
+    // Reschedule a second render that immediately returns the cached error
+    return {errorFromLastRender: error};
   }
   override componentDidUpdate() {
-    if (this.state.hasError) {
-      this.setState({hasError: false, error: null});
+    if (this.state.errorFromLastRender !== NO_ERROR_SENTINEL) {
+      // Reschedule a third render that immediately returns the cached error
```

Comment:
i don't get the comment here, seems backwards since we're clearing `errorFromLastRender`? wouldn't this _not_ immediately return the error?

Reply:
This wouldn't return `errorFromLastRender`, but it would find the correct one in `this.propsErrorMap`.
I had trouble figuring out another solution to render components exactly once (per entry in `sequentialRenders`) *and* clear the error-state to retry rendering when props change

Definitely not sold that this is the best way writing this and I'm probably missing a common pattern, any suggestions welcome.